server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -379,6 +430,45 @@ public KeywordFieldMapper build(MapperBuilderContext context) {
                 this
             );
         }
+
+        private FieldType fieldType(final IndexSortConfig indexSortConfig, final IndexMode indexMode, final String fullFieldName) {
+            if (shouldUseDocValuesSparseIndex(indexSortConfig, indexMode, fullFieldName)) {
+                return new FieldType(Defaults.FIELD_TYPE_WITH_SKIP_DOC_VALUES);
+            }
+            return new FieldType(Defaults.FIELD_TYPE);
+        }
+
+        private boolean shouldUseDocValuesSparseIndex(
+            final IndexSortConfig indexSortConfig,
+            final IndexMode indexMode,
+            final String fullFieldName
+        ) {
+            return (defaultIndexedAndDocValues() || isNotIndexedAndHasDocValues())
+                && isLogsDbMode(indexMode)
+                && isHostNameField(fullFieldName)
+                && isPrimarySortField(indexSortConfig);
+        }
+
+        private boolean isHostNameField(final String fullFieldName) {
```

Comment:
I think it is more readable if these sub conditions are used direct in the return statement (^)? The sub conditions aren't long and moving to private methods doesn't buy much?

Reply:
I separated them because there are many of those...but if you prefer I can extract them.